### DISCUSSION TITLE
Proving data streams?

### DISCUSSION THREAD

**Discussion by Scratch-net** [Consensys/gnark#982] [Q&A] [answered]
I see from the code that it's seems to be possible to prove dynamically sized data (compression for example). If it's true, what are design principles of such circuit and code overall? Let's say I have a hash or a cipher circuit and I'd like to be able to prove that I know a hash or a key to arbitrary sized data. Is it possible? Because that would definitely affect circuit size

**Comment by ivokub** [comment, marked as answer]
It is not possible. You would have to allocate the circuit for the possible maximum input size and then zero-pad the witness if the input witness is smaller.

**Reply by hussein-aitlahcen** [reply]
Maybe I can add that you can assert on a smaller input that the maximum input size. If you know that you have a maximum of 256 bit input, you can have an extra input that would tell you how many bits you need to constraint. An example is protobuf where decoding a varint is possible as it can't exceed 10 bytes.

**Reply by ivokub** [reply]
It is slightly more involved, we cannot use a witness variable as an argument to define how many bits we want to constraint. However, it is possible explicitly by constructing an upper bound (using bit composition or doublings) and then using `assert.IsLessEqualThan`.
